src/carcockpit/gltf_viewer_widget.cpp

```diff
@@ -0,0 +1,218 @@
+		this->fps = 0;
+	}
+
+	camera_position += (camera_attractor - camera_position) * fdt / camera_transition_duration;
+	ruis::vec3 remains = camera_attractor - camera_position;
+	ruis::real l2 = remains.x() * remains.x() + remains.y() * remains.y() + remains.z() * remains.z();
+	const ruis::real threshold = 0.00001;
+	if (l2 < threshold) {
+		camera_position = camera_attractor;
+	}
+
+	this->clear_cache();
+}
+
+void gltf_viewer_widget::toggle_camera(bool toggle)
+{
+	if (toggle)
+		camera_attractor = camera_position_top;
+	else
+		camera_attractor = camera_position_front;
+}
+
+void gltf_viewer_widget::set_normal_mapping(bool toggle)
+{
+	application::inst().shader_adv_v.set_normal_mapping(toggle);
+}
+
+constexpr float snap_speed = 1.07; // 1 is zero speed
+
+bool gltf_viewer_widget::on_mouse_button(const ruis::mouse_button_event& e)
+{
+	if (e.button == ruis::mouse_button::wheel_up) {
+		camera_attractor -= this->params.camera_target;
+		camera_attractor /= snap_speed;
+		camera_attractor += this->params.camera_target;
+
+		if (!this->params.smooth_navigation_zoom) {
+			camera_position -= this->params.camera_target;
+			camera_position /= snap_speed;
+			camera_position += this->params.camera_target;
+		}
+
+	} else if (e.button == ruis::mouse_button::wheel_down) {
+		camera_attractor -= this->params.camera_target;
+		camera_attractor *= snap_speed;
+		camera_attractor += this->params.camera_target;
+
+		if (!this->params.smooth_navigation_zoom) {
+			camera_position -= this->params.camera_target;
+			camera_position *= snap_speed;
+			camera_position += this->params.camera_target;
+		}
+
+	} else if (e.button == ruis::mouse_button::left) {
+		mouse_orbit = e.is_down;
+		if (e.is_down) {
+			mouse_changeview_start = e.pos;
+			camera_changeview_start = camera_position;
+		}
+	}
+	return true;
+}
+
+bool gltf_viewer_widget::on_mouse_move(const ruis::mouse_move_event& e)
+{
+	constexpr float mouse_orbit_speed_multiplier = 2.0f;
+
+	if (mouse_orbit) {
+		ruis::vec2 diff = e.pos - mouse_changeview_start;
+
+		ruis::vec3 cam2go = camera_changeview_start - this->params.camera_target;
+		ruis::vec3 axis = -cam2go.cross(ruis::vec3(0, 1, 0));
+		axis.normalize();
+
+		auto cam_norm = cam2go.normed();
+		float theta = ::acosf(cam_norm.y());
+		float dtheta = -diff.y() * mouse_orbit_speed_multiplier / (rect().d.y() + 1);
+
+		// NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+		float theta_lower_limit = static_cast<float>(M_PI_2) - params.orbit_angle_upper_limit + 0.00001f;
+		// NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+		float theta_upper_limit = params.orbit_angle_lower_limit + static_cast<float>(M_PI_2) - 0.00001f;
+
+		// restrict camera orbit angle to respect given range
+		if (theta + dtheta < theta_lower_limit)
+			dtheta = theta_lower_limit - theta;
+		if (theta + dtheta > theta_upper_limit)
+			dtheta = theta_upper_limit - theta;
+
+		ruis::quat q1, q2;
+		q1.set_rotation(axis.x(), axis.y(), axis.z(), dtheta);
```

Comment:
tehre is a `quaternion::set_rotation(axis_vector, angle)` overload, use it here. No need to pass axis components one by one.

Reply:
fixed